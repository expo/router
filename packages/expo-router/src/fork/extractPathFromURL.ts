import Constants, { ExecutionEnvironment } from "expo-constants";
import * as Linking from "expo-linking";

// This is only run on native.
export function extractExpoPathFromURL(url: string) {
  // Handle special URLs used in Expo Go: `/--/pathname` -> `pathname`
  if (Constants.executionEnvironment === ExecutionEnvironment.StoreClient) {
    const pathname = url.match(/exps?:\/\/.*?\/--\/(.*)/)?.[1];
    if (pathname) {
      return pathname;
    }
    // Fallback on default behavior
  }

  const res = Linking.parse(url);
  const qs = !res.queryParams
    ? ""
    : Object.entries(res.queryParams)
        .map(([k, v]) => `${k}=${v}`)
        .join("&");
  return (res.path || "") + (qs ? "?" + qs : "");
}
