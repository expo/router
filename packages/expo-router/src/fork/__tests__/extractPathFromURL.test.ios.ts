import Constants, { ExecutionEnvironment } from "expo-constants";

import { extractExpoPathFromURL } from "../extractPathFromURL";

describe(extractExpoPathFromURL, () => {
  const originalExecutionEnv = Constants.executionEnvironment;

  afterEach(() => {
    Constants.executionEnvironment = originalExecutionEnv;
  });

  test.each<string>([
    "exp://127.0.0.1:19000/",
    "exp://127.0.0.1:19000/--/test/path?query=param",
    "exp://127.0.0.1:19000?query=param",
    "exp://exp.host/@test/test/--/test/path?query=param",
    "exp://exp.host/@test/test/--/test/path",
    "exp://exp.host/@test/test/--/test/path/--/foobar",
    "https://example.com/test/path?query=param",
    "https://example.com/test/path",
    "https://example.com:8000/test/path",
    "https://example.com:8000/test/path+with+plus",
    "https://example.com/test/path?query=do+not+escape",
    "https://example.com/test/path?missingQueryValue=",
    "custom:///?shouldBeEscaped=x%252By%2540xxx.com",
    "custom:///test/path?foo=bar",
    "custom:///",
    "custom://",
    "custom://?hello=bar",
    "invalid",
  ])(`parses %p`, (url) => {
    Constants.executionEnvironment = ExecutionEnvironment.StoreClient;

    const res = extractExpoPathFromURL(url);
    expect(res).toMatchSnapshot();
    // Ensure the Expo Go handling never breaks
    expect(res).not.toMatch(/^--\//);
  });

  it(`only handles Expo Go URLs in Expo Go`, () => {
    Constants.executionEnvironment = ExecutionEnvironment.Bare;

    const res = extractExpoPathFromURL("exp://127.0.0.1:19000/--/test");
    expect(res).toEqual("--/test");
  });
});
