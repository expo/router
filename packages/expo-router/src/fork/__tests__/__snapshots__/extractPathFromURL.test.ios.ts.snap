// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractExpoPathFromURL parses "custom://" 1`] = `""`;

exports[`extractExpoPathFromURL parses "custom:///" 1`] = `""`;

exports[`extractExpoPathFromURL parses "custom:///?shouldBeEscaped=x%252By%2540xxx.com" 1`] = `"?shouldBeEscaped=x+y@xxx.com"`;

exports[`extractExpoPathFromURL parses "custom:///test/path?foo=bar" 1`] = `"test/path?foo=bar"`;

exports[`extractExpoPathFromURL parses "custom://?hello=bar" 1`] = `"?hello=bar"`;

exports[`extractExpoPathFromURL parses "exp://127.0.0.1:19000/" 1`] = `""`;

exports[`extractExpoPathFromURL parses "exp://127.0.0.1:19000/--/test/path?query=param" 1`] = `"test/path?query=param"`;

exports[`extractExpoPathFromURL parses "exp://127.0.0.1:19000?query=param" 1`] = `"?query=param"`;

exports[`extractExpoPathFromURL parses "exp://exp.host/@test/test/--/test/path" 1`] = `"test/path"`;

exports[`extractExpoPathFromURL parses "exp://exp.host/@test/test/--/test/path/--/foobar" 1`] = `"test/path/--/foobar"`;

exports[`extractExpoPathFromURL parses "exp://exp.host/@test/test/--/test/path?query=param" 1`] = `"test/path?query=param"`;

exports[`extractExpoPathFromURL parses "https://example.com/test/path" 1`] = `"test/path"`;

exports[`extractExpoPathFromURL parses "https://example.com/test/path?missingQueryValue=" 1`] = `"test/path?missingQueryValue="`;

exports[`extractExpoPathFromURL parses "https://example.com/test/path?query=do+not+escape" 1`] = `"test/path?query=do not escape"`;

exports[`extractExpoPathFromURL parses "https://example.com/test/path?query=param" 1`] = `"test/path?query=param"`;

exports[`extractExpoPathFromURL parses "https://example.com:8000/test/path" 1`] = `"test/path"`;

exports[`extractExpoPathFromURL parses "https://example.com:8000/test/path+with+plus" 1`] = `"with+plus"`;

exports[`extractExpoPathFromURL parses "invalid" 1`] = `"invalid"`;
